pkg/ipam/node.go

```diff
@@ -599,31 +596,84 @@ func (n *Node) maintainIPPool(ctx context.Context) error {
 
 	scopedLog := n.logger()
 
-	// Release excess addresses
-	if a.release != nil && len(a.release.IPsToRelease) > 0 {
-		var ipsToRelease []string
-		// Since, the default CRD sync interval is 15 secs, there might be new allocations that aren't reflected in
-		// the ciliumnode object yet. Verify against pod cache to check if any of the IPs are in use.
-		for _, ip := range a.release.IPsToRelease {
+	// Attempt to release if :
+	// * IP is still marked for release in the current iteration's a.release.IPsToRelease
+	// * Marked for release for more than excess-ip-release-delay seconds
+	// * IP is not in use by any pod according to pod informer cache
+	var ipsToRelease []string
+	if n.ipsMarkedForRelease != nil && len(n.ipsMarkedForRelease) > 0 {
+		if a.release == nil || len(a.release.IPsToRelease) == 0 {
+			n.ipsMarkedForRelease = make(map[string]time.Time)
```

Comment:
Doesn't this clear out `n.ipsMarkedForRelease` when we've just checked above that it has content?